Add method to dynamically add a deferred service.

tabennett opened the pull request:
This PR adds a method onto the application class that allows users to dynamically add items onto the application's deferredServices array (without having to reset the entire array).   

With this PR, users now have a way to dynamically add deferred service providers without having to hard code them inside the config/app.php providers array.

tabennett committed:
Add method to dynamically add a deferred service.

Added a simple method onto the application class that allow users to add
an item to the application's deferredServices array.

taylorotwell commented:
Won't you need to also pass the service provider? I think the service is the key and the provider is the value.

tabennett committed:
Make addDeferredServices accept an array.

The addDeferredServices method now accepts an array of values so that
multiple service => provider pairs can be added at a time.

tabennett commented:
Yeah, you're right man.  When I first looked at this I thought that the ```deferredServices``` var was just a list of service providers that would be deferred.  I see now that this isn't the case and what we're looking for here are key => value pairs of services and providers.  I've updated the method to accept and array of service => provider pairs so that multiple deferred services can be added at once (this seems like a better idea than forcing people to add them one at a time given that some people might have more than one to add).  

Playing around with this, I've was able to successfully dynamically add deferred services (from a customer service provider) as well override existing deferred services ('command.up' and 'command.down') with no problem.  If you need a test or two written for this just let me know.

crynobone commented:
I use a slightly more complex approach to deal with an package loading other deferred package. 

<https://github.com/orchestral/extension/blob/master/src/ProviderRepository.php#L40-L84>, and as for the `Application` <https://github.com/orchestral/foundation/blob/master/src/Application.php#L72-L80>.

For what I see based on your changes, you need to get all the `provides` information from the service provider and pass it to `Application`, it would be enough until the where package that you depend on suddenly change the `provides` list for any reason.

tabennett commented:
Yeah, I really don't want to extend Laravel's Illuminate\Foundation\Application though (if at all possible).  For what I'm trying to do it really just doesn't make much sense to do that.  I really just need to have setter/getters for the ```$deferredServices``` attribute.  In fact, maybe it might make sense to add a getter method for them to the container also (like what @crynobone is doing in his app class).  It would add a bit more flexibility for folks working with these lower level components, etc.  @crynobone I really like the implementation of your ProviderRepository class.

tabennett committed:
Add getter for the defferedService property.

Added a simple getter method for the Application's deferredServices
property.
